src/Framework/Analyzer/test/MinimalActions/DisallowReturningActionResultsFromMapMethodsTest.cs

```diff
@@ -0,0 +1,255 @@
+    }
+
+    [Fact]
+    public async Task MinimalAction_ReturningCustomIResult_Works()
+    {
+        // Arrange
+        var source = @"
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+var webApp = WebApplication.Create();
+webApp.MapGet(""/"", () => new CustomResult());
+
+class CustomResult : IResult
+{
+    public Task ExecuteAsync(HttpContext context) => Task.CompletedTask;
+}
+";
+        // Act
+        var diagnostics = await Runner.GetDiagnosticsAsync(source);
+
+        // Assert
+        Assert.Empty(diagnostics);
+    }
+
+    [Fact]
+    public async Task MinimalAction_ReturningIResultConditionallyWorks()
+    {
+        // Arrange
+        var source = @"
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+var webApp = WebApplication.Create();
+webApp.MapGet(""/"", (int id) =>
+{
+    if (id == 0)
+    {
+        return Results.NotFound();
+    }
+
+    return Results.Ok(""Here you go"");
+});
+
+";
+        // Act
+        var diagnostics = await Runner.GetDiagnosticsAsync(source);
+
+        // Assert
+        Assert.Empty(diagnostics);
+    }
+
+    [Fact]
+    public async Task MinimalAction_ReturningTypeThatImplementsIResultAndActionResultFails()
+    {
+        // Arrange
+        var source = @"
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+var webApp = WebApplication.Create();
+webApp.MapGet(""/"", () => new CustomResult());
+
+class CustomResult : IResult, IActionResult
+{
+    public Task ExecuteAsync(HttpContext context) => Task.CompletedTask;
+
+    public Task ExecuteResultAsync(ActionContext context) => Task.CompletedTask;
+}
+";
+        // Act
+        var diagnostics = await Runner.GetDiagnosticsAsync(source);
+
+        // Assert
+        Assert.Empty(diagnostics);
+    }
+
+    [Fact]
+    public async Task MinimalAction_ReturningActionResult_ProducesDiagnostics()
+    {
+        // Arrange
+        var source = TestSource.Read(@"
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+
+var webApp = WebApplication.Create();
+webApp.MapGet(""/"", () => /*MM*/new OkObjectResult(""cool story""));
+");
+        // Act
+        var diagnostics = await Runner.GetDiagnosticsAsync(source.Source);
+
+        // Assert
+        var diagnostic = Assert.Single(diagnostics);
+        Assert.Same(DiagnosticDescriptors.DoNotReturnActionResultsFromMapActions, diagnostic.Descriptor);
+        AnalyzerAssert.DiagnosticLocation(source.DefaultMarkerLocation, diagnostic.Location);
+        Assert.Equal("IActionResult instances should not be returned from a MapGet delegate parameter. Consider returning an equivalent result from Microsoft.AspNetCore.Http.Results.", diagnostic.GetMessage());
```

Comment:
We could go crazy here and make figure out the suggestion based on the context (you could even do a codefix).